# Interactive Selections

In [1]:
import altair as alt
from vega_datasets import data
import pandas as pd

# Cars

In [2]:
import altair as alt
from vega_datasets import data

cars = data.cars()
cars.rename(columns={"Miles_per_Gallon": "Miles per Gallon", "Weight_in_lbs": "Weight (lbs)"}, inplace=True)
# make the index into a column
cars.insert(0, 'id', cars.index)
cars.dropna(inplace=True)
cars.head()


id                       Name  Miles per Gallon  Cylinders  Displacement  \
0   0  chevrolet chevelle malibu              18.0          8         307.0   
1   1          buick skylark 320              15.0          8         350.0   
2   2         plymouth satellite              18.0          8         318.0   
3   3              amc rebel sst              16.0          8         304.0   
4   4                ford torino              17.0          8         302.0   

   Horsepower  Weight (lbs)  Acceleration       Year Origin  
0       130.0          3504          12.0 1970-01-01    USA  
1       165.0          3693          11.5 1970-01-01    USA  
2       150.0          3436          11.0 1970-01-01    USA  
3       150.0          3433          12.0 1970-01-01    USA  
4       140.0          3449          10.5 1970-01-01    USA

In [3]:
brush_selection = alt.selection_interval()

base_plot = alt.Chart(cars)


scatterplot = base_plot.mark_point().encode(
    x="Miles per Gallon:Q",
    y="Weight (lbs):Q",
    color=alt.condition(brush_selection, "Origin:N", alt.value("gray")),
    opacity=alt.condition(brush_selection, alt.value(0.7), alt.value(0.3))
    
).add_params(
    brush_selection
)

histogram = base_plot.mark_bar().encode(
    y="Origin:N",
    color="Origin:N",
    x="count():Q",
).transform_filter(
    brush_selection
)

scatterplot & histogram


alt.VConcatChart(...)

In [4]:
cars.to_csv("cars_clean.csv", index=False)

## Question 1: How many cars from Japan have a Miles Per Gallon value greater than 35

**Answer: 17**


In [5]:
q1_cars = cars[cars["Origin"] == "Japan"]
q1_cars = q1_cars[q1_cars["Miles per Gallon"] > 35]
q1_cars["Origin"].count()

17

## Question 2: Consider only cars that have a miles per gallon value greater than 30 AND weigh more than 2000 pounds. Which country produces the most and the least of these cars?

**Answer:**

Most: Japan (24) 
Least: Europe (11)

In [6]:
subset2 = cars[cars["Miles per Gallon"] > 30]
subset2 = subset2[subset2["Weight (lbs)"] > 2000]
subset2.groupby("Origin").count()

id  Name  Miles per Gallon  Cylinders  Displacement  Horsepower  \
Origin                                                                    
Europe  11    11                11         11            11          11   
Japan   24    24                24         24            24          24   
USA     17    17                17         17            17          17   

        Weight (lbs)  Acceleration  Year  
Origin                                    
Europe            11            11    11  
Japan             24            24    24  
USA               17            17    17

# Weather

In [7]:
weather = data.seattle_weather()
weather = pd.read_csv("https://raw.githubusercontent.com/vega/vega/main/docs/data/seattle-weather.csv")
weather = weather[weather["date"] > "2015-01-01"]

In [8]:
weather.rename(columns={"date": "Date", "temp_max": "Maximum Temperature", "weather": "Weather"}, inplace=True)
weather.reset_index(inplace=True)
weather.insert(0, 'id', weather.index)
weather.head(30)

id  index        Date  precipitation  Maximum Temperature  temp_min  wind  \
0    0   1097  2015-01-02            1.5                  5.6       0.0   2.3   
1    1   1098  2015-01-03            0.0                  5.0       1.7   1.7   
2    2   1099  2015-01-04           10.2                 10.6       3.3   4.5   
3    3   1100  2015-01-05            8.1                 12.2       9.4   6.4   
4    4   1101  2015-01-06            0.0                 12.2       6.1   1.3   
5    5   1102  2015-01-07            0.0                  7.8       5.6   1.6   
6    6   1103  2015-01-08            0.0                  7.8       1.7   2.6   
7    7   1104  2015-01-09            0.3                 10.0       3.3   0.6   
8    8   1105  2015-01-10            5.8                  7.8       6.1   0.5   
9    9   1106  2015-01-11            1.5                  9.4       7.2   1.1   
10  10   1107  2015-01-12            0.0                 11.1       4.4   1.6   
11  11   1108  2015-01-13            0.0                  9.4       2.8   2.7   
12  12   1109  2015-01-14            0.0                  6.1       0.6   2.8   
13  13   1110  2015-01-15            9.7                  7.8       1.1   3.2   
14  14   1111  2015-01-16            0.0                 11.7       5.6   4.5   
15  15   1112  2015-01-17           26.2                 13.3       3.3   2.8   
16  16   1113  2015-01-18           21.3                 13.9       7.2   6.6   
17  17   1114  2015-01-19            0.5                 10.0       6.1   2.8   
18  18   1115  2015-01-20            0.0                 10.0       3.3   3.0   
19  19   1116  2015-01-21            0.0                  7.2      -0.5   1.3   
20  20   1117  2015-01-22            0.8                  9.4       6.1   1.3   
21  21   1118  2015-01-23            5.8                 12.2       8.3   2.6   
22  22   1119  2015-01-24            0.5                 14.4      11.1   3.3   
23  23   1120  2015-01-25            0.0                 17.2       7.2   1.4   
24  24   1121  2015-01-26            0.0                 16.1       6.1   2.2   
25  25   1122  2015-01-27            0.8                 11.1       8.3   2.0   
26  26   1123  2015-01-28            0.0                 12.2       5.0   1.8   
27  27   1124  2015-01-29            0.0                 12.2       3.3   2.9   
28  28   1125  2015-01-30            0.0                  8.3       1.1   0.8   
29  29   1126  2015-01-31            0.0                  7.2       3.3   1.9   

   Weather  
0     rain  
1      fog  
2     rain  
3     rain  
4      fog  
5      fog  
6      fog  
7     rain  
8     rain  
9     rain  
10     fog  
11     fog  
12     fog  
13    rain  
14     fog  
15    rain  
16    rain  
17    rain  
18     fog  
19     fog  
20    rain  
21    rain  
22    rain  
23     fog  
24     fog  
25    rain  
26     fog  
27     sun  
28     fog  
29     fog

In [9]:
brush_selection = alt.selection_interval()

base_plot = alt.Chart(weather)


scatterplot = base_plot.mark_point().encode(
    x="Date:T",
    y="Maximum Temperature:Q",
    color=alt.condition(brush_selection, "Weather:N", alt.value("gray")),
    opacity=alt.condition(brush_selection, alt.value(0.7), alt.value(0.3))
    
).add_params(
    brush_selection
)

histogram = base_plot.mark_bar().encode(
    y="Weather:N",
    color="Weather:N",
    x="count():Q",
).transform_filter(
    brush_selection
)

scatterplot & histogram

alt.VConcatChart(...)

In [10]:
weather.to_csv("seattle_weather_clean.csv", index=False)


## Question 1: How many days from September to the end of the year is it sunny?

**Answer: 32**

In [11]:
q1_weather = weather[weather["Weather"] == "sun"]
q1_weather = q1_weather[q1_weather["Date"] > "2015-09-01"].reset_index()
q1_weather.count()


level_0                32
id                     32
index                  32
Date                   32
precipitation          32
Maximum Temperature    32
temp_min               32
wind                   32
Weather                32
dtype: int64

## Question 2: Consider only days between June and October with temperatures above 20 degrees. What are the most and least common weather types during those days? 

**Answer:**

Most common: sun (76)   
Least common: drizzle (6)  

In [12]:
q2_weather = weather[weather["Maximum Temperature"] > 20]
q2_weather = q2_weather[q2_weather["Date"] > "2015-05-31"]
q2_weather = q2_weather[q2_weather["Date"] < "2015-11-01"].reset_index()
q2_weather.groupby("Weather").count()

level_0  id  index  Date  precipitation  Maximum Temperature  \
Weather                                                                 
drizzle        6   6      6     6              6                    6   
fog            9   9      9     9              9                    9   
rain          13  13     13    13             13                   13   
sun           76  76     76    76             76                   76   

         temp_min  wind  
Weather                  
drizzle         6     6  
fog             9     9  
rain           13    13  
sun            76    76

# Penguins

In [13]:
penguins = pd.read_csv("penguins.csv")
penguins.dropna(inplace=True)
penguins.rename(columns={"body_mass_g": "Body Mass (g)", "flipper_length_mm": "Flipper Length (mm)", "species": "Species"}, inplace=True)
penguins.reset_index(inplace=True)
penguins.insert(0, 'id', penguins.index)
penguins.head()

id  index Species     island  bill_length_mm  bill_depth_mm  \
0   0      0  Adelie  Torgersen            39.1           18.7   
1   1      1  Adelie  Torgersen            39.5           17.4   
2   2      2  Adelie  Torgersen            40.3           18.0   
3   3      4  Adelie  Torgersen            36.7           19.3   
4   4      5  Adelie  Torgersen            39.3           20.6   

   Flipper Length (mm)  Body Mass (g)     sex  
0                181.0         3750.0    MALE  
1                186.0         3800.0  FEMALE  
2                195.0         3250.0  FEMALE  
3                193.0         3450.0  FEMALE  
4                190.0         3650.0    MALE

In [14]:
brush_selection = alt.selection_interval()

base_plot = alt.Chart(penguins)

scatterplot = base_plot.mark_point().encode(
    x=alt.X("Body Mass (g):Q").scale(zero=False),
    y=alt.Y("Flipper Length (mm):Q").scale(zero=False),
    color=alt.condition(brush_selection, "Species:N", alt.value("gray")),
    opacity=alt.condition(brush_selection, alt.value(0.7), alt.value(0.3))
    
).add_params(
    brush_selection
)

histogram = base_plot.mark_bar().encode(
    y="Species:N",
    color="Species:N",
    x="count():Q",
).transform_filter(
    brush_selection
)

scatterplot & histogram

alt.VConcatChart(...)

In [15]:
penguins.to_csv("penguin_clean.csv", index=False)

## Question 1: How many Gentoo penguins weigh less than 4.5k grams (g)?

**Answer: 15**

In [16]:
q1_penguins = penguins[penguins["Species"] == "Gentoo"]
q1_penguins = q1_penguins[q1_penguins["Body Mass (g)"] < 4500]
q1_penguins.count()

id                     15
index                  15
Species                15
island                 15
bill_length_mm         15
bill_depth_mm          15
Flipper Length (mm)    15
Body Mass (g)          15
sex                    15
dtype: int64

## Question 2: Consider only penguins that weigh between 4k and 5k grams (g) and have flipper lengths between 200 and 215mm. What is the most and least common type of penguin in this subset? 

**Answer:**  
Most Common: Gentoo  
Least Common: Adelie


In [17]:
q2_penguins = penguins[penguins["Body Mass (g)"] > 4000]
q2_penguins = q2_penguins[q2_penguins["Body Mass (g)"] < 5000]
q2_penguins = q2_penguins[q2_penguins["Flipper Length (mm)"] >= 200]
q2_penguins = q2_penguins[q2_penguins["Flipper Length (mm)"] <= 215]
q2_penguins.groupby("Species").count()

id  index  island  bill_length_mm  bill_depth_mm  \
Species                                                       
Adelie      4      4       4               4              4   
Chinstrap  12     12      12              12             12   
Gentoo     38     38      38              38             38   

           Flipper Length (mm)  Body Mass (g)  sex  
Species                                             
Adelie                       4              4    4  
Chinstrap                   12             12   12  
Gentoo                      38             38   38

# Movies

In [18]:
movies = data.movies()
movies = movies[movies["IMDB_Votes"] > 40000]
movies = movies[movies["US_Gross"] > 0]
movies = movies[movies["US_Gross"] <  490000000]
movies.reset_index(drop=True, inplace=True)
movies.insert(0, 'id', movies.index)
movies.head()


id                       Title    US_Gross  Worldwide_Gross  US_DVD_Sales  \
0   0       To Kill A Mockingbird  13129846.0       13129846.0           NaN   
1   1              Twelve Monkeys  57141459.0      168841459.0           NaN   
2   2       2001: A Space Odyssey  56700000.0       68700000.0           NaN   
3   3                   The Abyss  54243125.0       54243125.0           NaN   
4   4  Ace Ventura: Pet Detective  72217396.0      107217396.0           NaN   

   Production_Budget Release_Date MPAA_Rating  Running_Time_min  \
0          2000000.0  Dec 25 1962        None               NaN   
1         29000000.0  Dec 27 1995           R               NaN   
2         10500000.0  Apr 02 1968        None               NaN   
3         70000000.0  Aug 09 1989       PG-13               NaN   
4         12000000.0  Feb 04 1994       PG-13               NaN   

        Distributor               Source Major_Genre         Creative_Type  \
0         Universal                 None        None                  None   
1         Universal  Based on Short Film       Drama       Science Fiction   
2              None  Original Screenplay        None       Science Fiction   
3  20th Century Fox  Original Screenplay      Action       Science Fiction   
4      Warner Bros.  Original Screenplay      Comedy  Contemporary Fiction   

          Director  Rotten_Tomatoes_Rating  IMDB_Rating  IMDB_Votes  
0             None                    97.0          8.4     82786.0  
1    Terry Gilliam                     NaN          8.1    169858.0  
2  Stanley Kubrick                    96.0          8.4    160342.0  
3    James Cameron                    88.0          7.6     51018.0  
4      Tom Shadyac                    49.0          6.6     63543.0

In [19]:

movies = movies[["Title", "Rotten_Tomatoes_Rating", "IMDB_Rating", "Major_Genre", "US_Gross", "Production_Budget"]]
movies.dropna(inplace=True)
movies.head()


Title  Rotten_Tomatoes_Rating  IMDB_Rating  \
3                   The Abyss                    88.0          7.6   
4  Ace Ventura: Pet Detective                    49.0          6.6   
6                  Annie Hall                    98.0          8.2   
7              Apocalypse Now                    98.0          8.6   
8          Back to the Future                    96.0          8.4   

  Major_Genre     US_Gross  Production_Budget  
3      Action   54243125.0         70000000.0  
4      Comedy   72217396.0         12000000.0  
6      Comedy   38251425.0          4000000.0  
7      Action   78800000.0         31500000.0  
8   Adventure  210609762.0         19000000.0

In [20]:
genres_to_keep = ['Action', 'Adventure', 'Comedy', 'Drama']
movies = movies[movies["Major_Genre"].isin(genres_to_keep)]
movies.rename(columns={"Major_Genre": "Genre", "US_Gross": "Gross Income", "Production_Budget": "Production Budget"}, inplace=True)
movies.head()

Title  Rotten_Tomatoes_Rating  IMDB_Rating      Genre  \
3                   The Abyss                    88.0          7.6     Action   
4  Ace Ventura: Pet Detective                    49.0          6.6     Comedy   
6                  Annie Hall                    98.0          8.2     Comedy   
7              Apocalypse Now                    98.0          8.6     Action   
8          Back to the Future                    96.0          8.4  Adventure   

   Gross Income  Production Budget  
3    54243125.0         70000000.0  
4    72217396.0         12000000.0  
6    38251425.0          4000000.0  
7    78800000.0         31500000.0  
8   210609762.0         19000000.0

In [21]:
base_plot = alt.Chart(movies)

scatterplot = base_plot.mark_point().encode(
    x=alt.X("Production Budget:Q").scale(zero=False),
    y=alt.Y("Gross Income:Q").scale(zero=False),
    color=alt.condition(brush_selection, "Genre:N", alt.value("gray")),
    opacity=alt.condition(brush_selection, alt.value(0.7), alt.value(0.3))
).add_params(
    brush_selection
)

histogram = base_plot.mark_bar().encode(
    y="Genre:N",
    color="Genre:N",
    x="count():Q",
).transform_filter(
    brush_selection
)

scatterplot & histogram

alt.VConcatChart(...)

In [22]:
movies.to_csv("movies_clean.csv")

## Questions and Answers

### Question 1: How many Adventure Movies have a production budget value greater than 100 million? 

**Answer: 28**

In [23]:
adventures = movies[movies["Genre"] == "Adventure"].reset_index(drop=True)
adventures.head(70)


Title  Rotten_Tomatoes_Rating  \
0                      Back to the Future                    96.0   
1              Back to the Future Part II                    64.0   
2             Back to the Future Part III                    71.0   
3                                 Ben-Hur                    91.0   
4      Close Encounters of the Third Kind                    95.0   
..                                    ...                     ...   
65                           Spider-Man 3                    63.0   
66                             Spider-Man                    89.0   
67                       Superman Returns                    76.0   
68  The League of Extraordinary Gentlemen                    17.0   
69                 The Day After Tomorrow                    46.0   

    IMDB_Rating      Genre  Gross Income  Production Budget  
0           8.4  Adventure   210609762.0         19000000.0  
1           7.5  Adventure   118450002.0         40000000.0  
2           7.1  Adventure    87666629.0         40000000.0  
3           8.2  Adventure    73000000.0         15000000.0  
4           7.8  Adventure   166000000.0         20000000.0  
..          ...        ...           ...                ...  
65          6.4  Adventure   336530303.0        258000000.0  
66          7.4  Adventure   403706375.0        139000000.0  
67          6.6  Adventure   200120000.0        232000000.0  
68          5.5  Adventure    66465204.0         78000000.0  
69          6.3  Adventure   186740799.0        125000000.0  

[70 rows x 6 columns]

In [24]:
big_budget_adventures = adventures[adventures["Production Budget"] > 100000000]
count = big_budget_adventures["Title"].count()
print(f"There are {count} big budget adventures")

There are 28 big budget adventures


## Question 2: Consider only movies that had a production budget greater than 100 million and made less than 150  Million in Gross income. Which genres had the most and the least movies in this group of movies?

**Answer:**  
Most: Action (19)  
Least: Comedy (1)

In [25]:
q2_movies = movies[movies["Production Budget"] > 100000000]
q2_movies = q2_movies[q2_movies["Gross Income"] < 150000000]
q2_movies.groupby("Genre").count()

Title  Rotten_Tomatoes_Rating  IMDB_Rating  Gross Income  \
Genre                                                                 
Action        19                      19           19            19   
Adventure      9                       9            9             9   
Comedy         1                       1            1             1   
Drama          4                       4            4             4   

           Production Budget  
Genre                         
Action                    19  
Adventure                  9  
Comedy                     1  
Drama                      4